crypto/schnorr.c

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright (c) 2021 The Bitcoin ABC developers
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "schnorr.h"
+#include "hmac_drbg.h"
+#include "memzero.h"
+#include "rfc6979.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one = 0;
+  uint32_t r = 0;
+  bignum256 n = {0}, k = {0}, temp = {0};
+
+  assert(!bn_is_zero(_k) && bn_is_odd(_k));
+
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
+
+  while (!bn_is_zero(&n)) {
+    while (bn_is_even(&n)) {
+      bn_rshift(&n);
+      // jacobi(2 * n, k) = jacobi(n, k) if k = 1 (mod 8) or k = 7 (mod 8)
+      // jacobi(2 * n, k) = -jacobi(n, k) if k = 3 (mod 8) or k = 5 (mod 8)
+      r = k.val[0] & 0x07;
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_copy(&n, &temp);
+    bn_copy(&k, &n);
+    bn_copy(&temp, &k);
+
+    // jacobi(n, k) = jacobi(k, n) if k = n = 1 (mod 4)
+    // jacobi(n, k) = -jacobi(k, n) if k = n = 3 (mod 4)
+    t ^= ((n.val[0] & k.val[0] & 3) == 3);
+
+    bn_mod(&n, &k);
```

Comment:
I thought I commented it too, but I probably didn't...

The problem is that the function `bn_fast_mod(n, k)` you want to use works only for `2^256 - 2^224 <= k <= 2^256`. So we don't have a function for general reduction.

> so you might as well replace this line with bn_subtract(&n, &k, &n),

You can't, try `jacobi(7, 1)`.

I think the algorithm works correctly. My experimental results show that it needs about three times more iterations in the outer loop for 256-bit numbers. If the quadratic reciprocity is used only if `n < k` (so it either subtracts or uses the quadratic reciprocity) the algorithm needs only two times more iterations. That's not so bad.

Reply:
I never suggested using `bn_fast_mod()`. My point is that the `bn_mod()` here is highly misleading, because it will result in a different flow than one would expect and even when one realizes this, it's still far from obvious that the return value will be correct.

Reply:
Fixed in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee. @onvej-sl please review and resolve.

Reply:
> I never suggested using bn_fast_mod().

And I didn't mean it's you how want to use it. :)

Reply:
Looks good!